gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -252,27 +344,46 @@ private WriteObjectResponse doResumableUpload() throws IOException {
                 resourceId, responseObserver.uploadId));
       }
 
+      long numberOfBufferedRequests = channelOptions.getNumberOfBufferedRequests();
+      long freeBufferThreshold =
+          Math.max(numberOfBufferedRequests - 5, numberOfBufferedRequests / 2);
```

Comment:
This is better modeled as int IMO, with decimal and negative values possibly causing bugging buffer clean ups